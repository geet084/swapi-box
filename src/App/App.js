import React, { Component } from 'react';
import './App.scss';
import ScrollBox from '../ScrollBox/ScrollBox'
import Favorites from '../Favorites/Favorites'
import Nav from '../Nav/Nav'
import Display from '../Display/Display'

export default class App extends Component {
  constructor() {
    super();
    this.state = {
      movie: {},
      people: []
    }
  }

  findRandomMovie = (movies) => {
    let maxNum = Math.floor(movies.length);
    let randomNumber = Math.floor(Math.random() * maxNum)

    return movies[randomNumber];
  }

  componentDidMount = () => {
    fetch('https://swapi.co/api/films/')
      .then(response => response.json())
      .then(movies => this.findRandomMovie(movies.results))
      .then(movie => this.setState({ movie }))
  }

  fetchPeople = () => {
    let { people } = this.state;
    let pageNum = (people.length % 9) + 1;
    if (people.length < 87) {
      const url = 'https://swapi.co/api/people/?page='
    
      fetch(url + pageNum)
      .then(response => response.json())
      .then(peopleList => peopleList.results)
      .then(peopleWorlds => this.fetchWorldInfo(peopleWorlds))
      .then(peopleSpecies => this.fetchSpeciesInfo(peopleSpecies))
      .then(newPeople => {
        return this.setState({
          people: [...people, ...newPeople]
        })
      })
    }
  }

  fetchSpeciesInfo = (peeps) => {
    let updatedSpecies = peeps.map(peep => {
      if (peep.species.length === 0) {
        return ({ ...peep, spec: 'unknown' })
      } else {
        return fetch(peep.species)
          .then(response => response.json())
          .then(spec => ({ ...peep, spec: spec.name }))
      }
    })
    return Promise.all(updatedSpecies);
  }

  fetchWorldInfo = (people) => {
   let updatedPeople = people.map(person => {
      return fetch(person.homeworld)
        .then(response => response.json())
        .then(world => {
          let { name, population } = world;
          return ({...person, world: name, population})
        })
   })
    return Promise.all(updatedPeople);

  }

  render() {
    let { movie, people } = this.state;

    if (movie !== {}) {
      return (
        <main className="App">
          <ScrollBox movie={movie.opening_crawl} />
          <p>{movie.title}</p>
          <Favorites />
          <Nav showMore={this.fetchPeople} />
          <Display people={people} />
        </main>
      );
    } else {
      return <div>nothing to show</div>
    }
  }
}