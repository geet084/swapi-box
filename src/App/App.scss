body {
  background-image: url('../images/sw-bg.png')
}

.navigation {
  display: flex;
  justify-content: space-around;
  height: 125px;
  background-image: url('../images/saber.jpg');
  background-size: 100vw auto;
  background-position-y: center; 
}

.text {
  color: white;
  margin: 0;
}