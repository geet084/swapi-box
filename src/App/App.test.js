import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme'

describe('App', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(
      <App  />
    )
  })

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  it('should properly match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it.skip('should have sad path snapshot', () => {
    wrapper.setState({movie: {}})
    expect(wrapper).toMatchSnapshot()
  })

  it('should have a default state', () => {
    const expected = {
      url: 'https://swapi.co/api/',
      movie: {},
      activeChoice: '',
      people: [],
      planets: [],
      vehicles: []
    }
    expect(wrapper.state()).toEqual(expected)
  })

  it.skip('should find random movie', () => {

  })

  it.skip('should mount', () => {

  })

  it.skip('getMoreInfo', () => {

  })

  it.skip('fetchSpeciesInfo', () => {

  })

  it.skip('fetchWorldInfo', () => {

  })

  it.skip('fetchResidentsInfo', () => {

  })

  it.skip('fetchAPIInfo', () => {

  })
})
