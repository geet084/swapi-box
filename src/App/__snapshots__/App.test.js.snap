// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should have sad path snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <scrollBox />,
        <p
          className="text"
        />,
        <nav
          className="navigation"
        >
          <Favorites
            showFavorites={[Function]}
          />
          <Nav
            fetchInfo={[Function]}
          />
        </nav>,
        <Display
          activeChoice=""
          choice=""
          movie={Object {}}
          people={Array []}
          planets={Array []}
          url="https://swapi.co/api/"
          vehicles={Array []}
        />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "crawlText": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "text",
        },
        "ref": null,
        "rendered": null,
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Favorites
              showFavorites={[Function]}
            />,
            <Nav
              fetchInfo={[Function]}
            />,
          ],
          "className": "navigation",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "showFavorites": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "fetchInfo": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "nav",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "activeChoice": "",
          "choice": "",
          "movie": Object {},
          "people": Array [],
          "planets": Array [],
          "url": "https://swapi.co/api/",
          "vehicles": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "main",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <scrollBox />,
          <p
            className="text"
          />,
          <nav
            className="navigation"
          >
            <Favorites
              showFavorites={[Function]}
            />
            <Nav
              fetchInfo={[Function]}
            />
          </nav>,
          <Display
            activeChoice=""
            choice=""
            movie={Object {}}
            people={Array []}
            planets={Array []}
            url="https://swapi.co/api/"
            vehicles={Array []}
          />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "crawlText": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "text",
          },
          "ref": null,
          "rendered": null,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Favorites
                showFavorites={[Function]}
              />,
              <Nav
                fetchInfo={[Function]}
              />,
            ],
            "className": "navigation",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "showFavorites": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "fetchInfo": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "nav",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "activeChoice": "",
            "choice": "",
            "movie": Object {},
            "people": Array [],
            "planets": Array [],
            "url": "https://swapi.co/api/",
            "vehicles": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "main",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`App should properly match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <scrollBox />,
        <p
          className="text"
        />,
        <nav
          className="navigation"
        >
          <Favorites
            showFavorites={[Function]}
          />
          <Nav
            fetchInfo={[Function]}
          />
        </nav>,
        <Display
          activeChoice=""
          choice=""
          movie={Object {}}
          people={Array []}
          planets={Array []}
          url="https://swapi.co/api/"
          vehicles={Array []}
        />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "crawlText": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "text",
        },
        "ref": null,
        "rendered": null,
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Favorites
              showFavorites={[Function]}
            />,
            <Nav
              fetchInfo={[Function]}
            />,
          ],
          "className": "navigation",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "showFavorites": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "fetchInfo": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "nav",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "activeChoice": "",
          "choice": "",
          "movie": Object {},
          "people": Array [],
          "planets": Array [],
          "url": "https://swapi.co/api/",
          "vehicles": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "main",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <scrollBox />,
          <p
            className="text"
          />,
          <nav
            className="navigation"
          >
            <Favorites
              showFavorites={[Function]}
            />
            <Nav
              fetchInfo={[Function]}
            />
          </nav>,
          <Display
            activeChoice=""
            choice=""
            movie={Object {}}
            people={Array []}
            planets={Array []}
            url="https://swapi.co/api/"
            vehicles={Array []}
          />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "crawlText": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "text",
          },
          "ref": null,
          "rendered": null,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Favorites
                showFavorites={[Function]}
              />,
              <Nav
                fetchInfo={[Function]}
              />,
            ],
            "className": "navigation",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "showFavorites": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "fetchInfo": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "nav",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "activeChoice": "",
            "choice": "",
            "movie": Object {},
            "people": Array [],
            "planets": Array [],
            "url": "https://swapi.co/api/",
            "vehicles": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "main",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
